jvb/src/main/kotlin/org/jitsi/videobridge/relay/Relay.kt

```diff
@@ -0,0 +1,549 @@
+            bridgeJitterStats.packetSent(packetInfo)
+        } else if (packetInfo.packet.looksLikeRtcp()) {
+            Endpoint.rtcpPacketDelayStats.addPacket(packetInfo)
+        }
+         */
+
+        packetInfo.sent()
+        /* TODO
+        if (timelineLogger.isTraceEnabled && Endpoint.logTimeline()) {
+            timelineLogger.trace { packetInfo.timeline.toString() }
+        }
+         */
+        iceTransport.send(packetInfo.packet.buffer, packetInfo.packet.offset, packetInfo.packet.length)
+        ByteBufferPool.returnBuffer(packetInfo.packet.buffer)
+        return true
+    }
+
+    /**
+     * Sends a specific msg to this endpoint over its bridge channel
+     */
+    fun sendMessage(msg: BridgeChannelMessage) = _messageTransport.sendMessage(msg)
+
+    fun relayMessageTransportConnected() {
+        relayedEndpoints.values.forEach { e -> e.relayMessageTransportConnected() }
+    }
+
+    fun addRemoteEndpoint(
+        id: String,
+        statsId: String?,
+        audioSources: Collection<AudioSourceDesc>,
+        videoSources: Collection<MediaSourceDesc>
+    ) {
+        val ep: RelayedEndpoint
+        synchronized(endpointsLock) {
+            if (relayedEndpoints.containsKey(id)) {
+                logger.warn("Relay already contains remote endpoint with ID $id")
+                updateRemoteEndpoint(id, audioSources, videoSources)
+                return
+            }
+            ep = RelayedEndpoint(conference, this, id, logger)
+            ep.statsId = statsId
+            ep.audioSources = audioSources.toTypedArray()
+            ep.mediaSources = videoSources.toTypedArray()
+
+            relayedEndpoints[id] = ep
+
+            ep.ssrcs.forEach { ssrc -> endpointsBySsrc[ssrc] = ep }
+        }
+
+        conference.addEndpoints(setOf(ep))
+        updateTransceiverSources()
+    }
+
+    fun updateRemoteEndpoint(
+        id: String,
+        audioSources: Collection<AudioSourceDesc>,
+        videoSources: Collection<MediaSourceDesc>
+    ) {
+        val ep: RelayedEndpoint
+        synchronized(endpointsLock) {
+            ep = relayedEndpoints[id] ?: run {
+                logger.warn("Endpoint with ID $id not found in relay")
+                return
+            }
+            val oldSsrcs = ep.ssrcs
+
+            ep.audioSources = audioSources.toTypedArray()
+            ep.mediaSources = videoSources.toTypedArray()
+
+            val newSsrcs = ep.ssrcs
+            val removedSsrcs = oldSsrcs.minus(newSsrcs)
+            val addedSsrcs = newSsrcs.minus(oldSsrcs)
+
+            endpointsBySsrc.keys.removeAll(removedSsrcs)
+            addedSsrcs.forEach { ssrc -> endpointsBySsrc[ssrc] = ep }
+        }
+        updateTransceiverSources()
+    }
+
+    fun removeRemoteEndpoint(id: String) {
+        val ep: RelayedEndpoint?
+        synchronized(endpointsLock) {
+            ep = relayedEndpoints.remove(id)
+            if (ep != null) {
+                endpointsBySsrc.keys.removeAll(ep.ssrcs)
```

Comment:
I don't think it's possible for the endpoint's SSRC and the map to get out of sync, but maybe cleaning by the endpoint would make it more obvious? `endpointsBySsrc.entries.removeIf { it.value == ep }`

Reply:
That iterates over all the elements in the `endpointsBySsrc` map, though, so it'll be slower for large conferences.